### PR TITLE
Version for VS2019 using both CLANG and MSVC

### PR DESCRIPTION
**Pull request by DBJDBJ**
Builds and runs using both MSVC and CLANG 8.0.1 coming with Visual Studio 2019

### CODE REVIEW

File: test/windows_branch_readme.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    // output the pointer
+    pprintf("{}", message_) ;
+#endif
+```
+
+#### Experimental
+
+Currently the experimental subfolder contains `ppf_arg`. Schema that might be used to somewhat serve the users wanting to handle their user defined types.
+
+The purpose of the experimental stuff. is to be used for discussions. Will be eihter incorporated or erased.
```
Comment (tfc) [inline test/windows_branch_readme.md]: experimental stuff should go to an experimental branch. This branch might begin on this one.

File: include/pprintpp/typelist.hpp [hunk 2]
```diff
@@ -21,91 +23,98 @@
...
+		struct append<null_t, null_t> { using type = null_t; };
+		template <class T>
+		struct append<null_t, T> { using type = make_t<T>; };
+		template <class Head, class T>
+		struct append<null_t, tl<Head, T>> { using type = tl<Head, T>; };
+		template <class Head, class Tail, class T>
+		struct append<tl<Head, Tail>, T>
+		{
+			using type = tl<Head, typename append<Tail, T>::type>;
+		};
```
Comment (tfc) [inline include/pprintpp/typelist.hpp]: We should both adhere to the same coding style. I don't like that it's not immediately visible what was changed for the windows version because of additional style changes.

File: include/pprintpp/typelist.hpp [hunk 3]
```diff
@@ -21,91 +23,98 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-#pragma once
 
-namespace typelist {
+namespace jacek_galowicz {
```
Comment (tfc) [inline include/pprintpp/typelist.hpp]: wrapping stuff into namespaces with our names is not an advantage to the user, and i think people also don't really care who wrote which part. 

Let us please not add extra nestings

File: include/pprintpp/pprintpp.hpp [hunk 4]
```diff
@@ -21,159 +24,221 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-#pragma once
```
Comment (tfc) [inline include/pprintpp/pprintpp.hpp]: why did you remove the `pragma once` in favor of the `ifndef-define-endif`?

File: include/pprintpp/pprintpp.hpp [hunk 5]
```diff
@@ -21,159 +24,221 @@
...
+
+// DBJ -- to print 'char *' as a string is standard behaviour
+// #define PPRINTPP_STANDARD_CHAR_PTR 1
+
+#ifndef PPRINTPP_AVOID_STL
+#pragma message("PPRINTPP *using* STD lib and *throwning* exceptions\n\n")
```
Comment (tfc) [inline include/pprintpp/pprintpp.hpp]: i would prefer the compiler not cluttering the build output with non-critical messages.

File: include/pprintpp/pprintpp.hpp [hunk 6]
```diff
@@ -21,159 +24,221 @@
...
-template <typename ... Ts>
-make_t<Ts...> tie_types(Ts...);
-
-#define AUTOFORMAT(s, ...) \
-    ({ \
-        struct strprov { static constexpr const char * str() { return static_cast<const char*>(s); } }; \
+namespace jacek_galowicz {
+	namespace pprintpp {
+
+		constexpr auto VERSION = "0.2.0";
```
Comment (tfc) [inline include/pprintpp/pprintpp.hpp]: why did you add a version tag? Where and why would you use it?

File: test/windows_branch_readme.md (hunk 1 above)
Comment (DBJDBJ) [reply test/windows_branch_readme.md]: Will do, thanks

File: include/pprintpp/typelist.hpp (hunk 3 above)
Comment (DBJDBJ) [reply include/pprintpp/typelist.hpp]: namespace 'typelist' is very likely already in existence somewhere in the world
namespace 'jacek_galowicz' is very likely not
also this is completely transparent to the user
this is what I care for not "who wrote which" part ...

Comment (tfc) [reply include/pprintpp/typelist.hpp]: totally makes sense. how about having a general `pprintpp` namespace and `typelist` is part of just that? so we structure this by library structure and not by authorship?

Comment (DBJDBJ) [reply include/pprintpp/typelist.hpp]: let me do that ...

File: include/pprintpp/pprintpp.hpp [hunk 7]
```diff
@@ -24,8 +24,21 @@
  * SOFTWARE.
  */
 
-// DBJ -- to print 'char *' as a string is standard behaviour
-// #define PPRINTPP_STANDARD_CHAR_PTR 1
+/*
+GCC 5.5.0 is chocking on this one perhaps?
```
Comment (tfc) [inline include/pprintpp/pprintpp.hpp]: Where did you find such an old GCC version? Who uses that nowadays?

Comment (DBJDBJ) [reply include/pprintpp/pprintpp.hpp]: Was wondering myself. I have to say I deciphered that from failed CI build ... let me see again ...

Comment (DBJDBJ) [reply include/pprintpp/pprintpp.hpp]: Looked again, is that it?
```
$ cmake -DCMAKE_BUILD_TYPE=Release
-- The CXX compiler identification is GNU 5.5.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/travis/build/tfc/pprintpp
```

**Comment by DBJDBJ** [comment]
Travis fails on linking now ...

**Comment by DBJDBJ** [comment]
Hey TFC, 

I think you have here quite clever disruptor people will use. On windows too. In an unlimited fashion.

Currently, this is Nix only, char only lib.  And it works fine. 
But, I am looking for C++17/C++20 lib, which will also support char8_t, char16_t, char32_t, and wchar_t. 

I think you should decouple headers from test projects. Also for them to be easily used as git submodules. Nothing wrong with a multi-project approach. Small number of small projects. Each with a simple Travis set up for a specific compiler platform combination. 

What do you think?

**Comment by DBJDBJ** [comment]
Hey @tfc  can we resolve this pull request please? CI seems to be not configured properly. I would like to help but I have not experience whatsoever with CI. I blame it on Visual Studio.

**Comment by tfc** [comment]
Hey @DBJDBJ, i have limited availability these days, but i will have a thorough look at your PR in the next days.

The CI thing however is something we need to get under control. I am not going to merge anything that is not covered by a running CI, as this opens the gate to bitrotting.

**Comment by DBJDBJ** [comment]
You are right @tfc. CI is important. Unfortunately, it seems you do not have one configured on this project.

**Comment by tfc** [comment]
@DBJDBJ i have none configured on github but i have a private hydra server that runs this thing against different compilers - which will not work with windows environments. But you have full power of your own repository, so i suggest you find something that works that i could then also enable for this repository.

**Comment by DBJDBJ** [comment]
@tfc , let me be productive

1. me form project A on my side with pprintfpp headers only (so it can be used as github submodule)
2. me form project B in my side where I include A as a submodule and configure CIRCLE CI of the . Project B is windows only 

Same as you I am rather pressed for time. And I am not qualified to maintain multi OS multi compiler CI.

I hope you will agree ... Kind regards
